server/metadata/tenant.go

```diff
@@ -192,10 +192,34 @@ func (m *TenantManager) CreateTenant(ctx context.Context, tx transaction.Tx, nam
 	return namespace, nil
 }
 
-func (m *TenantManager) GetNamespace(namespaceName string) Namespace {
+func (m *TenantManager) GetNamespace(ctx context.Context, namespaceName string, txMgr *transaction.Manager) Namespace {
 	m.RLock()
-	defer m.RUnlock()
 	if tenant, found := m.tenants[namespaceName]; found {
+		defer m.RUnlock()
+		return tenant.namespace
+	}
+	m.RUnlock()
+
+	// this will never create new namespace
+	// this is for reading namespaces from storage into cache
+	tx, err := txMgr.StartTx(ctx)
+	if err != nil {
+		return nil
+	}
+
+	namespaces, err := m.encoder.GetNamespaces(ctx, tx)
+	err = tx.Commit(ctx)
+	if err != nil {
+		return nil
+	}
+	if id, ok := namespaces[namespaceName]; ok {
+		tenant, err := m.CreateOrGetTenant(ctx, txMgr, NewTenantNamespace(namespaceName, id))
```

Comment:
You are already starting a transaction not helpful to call `CreateOrGetTenant` because that method is starting a transaction as well.

Reply:
removed the call to createOrGetTenant